IOS-6286 Cardano tokens

ba1ashov opened the pull request:
- Убрана фильтрация `UTXO` с токенами в `Cardano`
- Добавлен `CardanoWithdrawalRestrictable` для отображения уведомлений в send
- Убрано лишнее перемапливание `Decimal` -> `UInt64`  в `Cardano`
- Переименовано `WithdrawalSuggestion` -> `WithdrawalNotification`

ba1ashov committed:
IOS-6286 Cardano tokens support

IOS-6286 Update Cardano token ValidationErrors

IOS-6286 Cardano tokens final

IOS-6286 Update DustRestrictable

IOS-6286 Add CardanoTokenWithdrawalRestrictable

IOS-6286 Update CardanoWalletManager.swift

Balashov152 commented on BlockchainSdk/Blockchains/Cardano/Adalite/AdaliteNetworkProvider.swift:
```diff
@@ -35,6 +35,12 @@ class AdaliteNetworkProvider: CardanoNetworkProvider {
             .mapNotEmptyString()
             .eraseError()
+            .map { hash in
+                // Remove \" from the string
+                // Because in response the hash is equal `"239392bc6c6bd354d5a95b77f799b61ac056c65566a10bb50033a7c9deabfe42"`
```
Не работала ссылка на explorer из за лишних кавычек

ba1ashov committed:
IOS-6286 Rename to `CardanoWithdrawalRestrictable`

ba1ashov committed:
IOS-6286 Fix review

IOS-6286 Update RosettaNetworkProvider.swift

IOS-6286 Update project.pbxproj

IOS-6286 Update project.pbxproj

ba1ashov committed:
IOS-6286 Update CardanoTransactionBuilder.swift

ba1ashov committed:
IOS-6286 WithdrawalSuggestion -> WithdrawalNotification

ba1ashov committed:
Merge branch 'develop' into IOS-6286-Cardano-tokens

tureck1y commented on BlockchainSdk/Common/Amount.swift:
```diff
@@ -68,7 +68,11 @@ public struct Amount: CustomStringConvertible, Hashable, Comparable {
     }
-    
+
+    public var decimalValue: Decimal {
+        pow(10, decimals)
+    }
```
Я бы эту штуку сюда не добавлял. Сейчас это лежит в блокчейне, что тоже не очень. По хорошему нужна утилита-конвертер.

1) Нам надо в целом переходить на целые числа, тогда это не понадобится
2) Смысл этого значения был просто в значении делителя для конвертами, а тут кажется что это значение amount в Decimal

Balashov152 replied:
Вообще да, согласен
надо переходить на BigInt в Amount. А в decimal через отдельный конвертер преобразовывать только для UI
Убрал ✅

ba1ashov committed:
IOS-6286 Fix review

Balashov152 commented on BlockchainSdk/Blockchains/Cardano/CardanoWalletManager.swift:
```diff
@@ -102,36 +104,151 @@ extension CardanoWalletManager: TransactionSender {
+            return
+        }
+
+        let minChange = try minChange(exclude: amount.type)
```
Убрал исключение в этом случае ✅

Balashov152 commented on BlockchainSdk/Blockchains/Cardano/CardanoWalletManager.swift:
```diff
@@ -102,36 +104,151 @@ extension CardanoWalletManager: TransactionSender {
+            throw ValidationError.balanceNotFound
+        }
+
+        let minChange = try minChange(exclude: amount.type)
```
Да, согласен для коина вообще это не нужно ✅

Balashov152 commented on BlockchainSdk/Blockchains/Cardano/CardanoWalletManager.swift:
```diff
@@ -102,36 +104,151 @@ extension CardanoWalletManager: TransactionSender {
+
+extension CardanoWalletManager: CardanoWithdrawalRestrictable {
+    func validateCardanoWithdrawal(amount: Amount, fee: Amount) throws {
+        let hasAnotherTokenWithBalance = wallet.amounts
```
Сделал по другому не много, этот флаг нужен если только весь токен отправляем ✅

ba1ashov committed:
IOS-6286 Update `hasAnotherTokenWithBalance`  flag

Balashov152 commented on BlockchainSdk/Blockchains/Cardano/CardanoWalletManager.swift:
```diff
@@ -102,36 +104,151 @@ extension CardanoWalletManager: TransactionSender {
+        }
+    }
+
+    private func validateCardanoCoinWithdrawal(amount: Amount, fee: Amount, hasAnotherTokenWithBalance: Bool) throws {
```
Сделал более просто 
`CardanoTransferRestrictable` / `validateCardanoTransfer`

dbaturin replied:
Публичную функцию переименовал, а приватные остались с "Withdrawal". Немного странно, хоть и не пренципиально

ba1ashov committed:
IOS-6286 Rename CardanoWithdrawalRestrictable -> CardanoTransferRestrictable

dbaturin commented on BlockchainSdk/Blockchains/Cardano/CardanoWalletManager.swift:
```diff
@@ -224,13 +212,20 @@ extension CardanoWalletManager: CardanoWithdrawalRestrictable {
 
-        let minChange = try minChange(exclude: amount.type)
-        var change = adaBalance - minAdaDecimal
+        let minChange = try minChange()
```
Почему здесь не исключаем текущий токен, если его баланса не осталось? Ведь мы дойдём сюда, если есть другие токены.

dbaturin commented on BlockchainSdk/Blockchains/Cardano/CardanoWalletManager.swift:
```diff
@@ -102,36 +104,146 @@ extension CardanoWalletManager: TransactionSender {
+            return
+        }
+
+        if hasAnotherTokenWithBalance {
+            throw ValidationError.cardanoHasTokens(minimumAmount: minChange)
```
А здесь мы кинем экспепшн наоборот, только если есть другие токены.

Balashov152 replied:
Согласен, упростил, теперь везде `ValidationError.cardanoInsufficientBalanceToSendToken` ✅

dbaturin commented on BlockchainSdk/Blockchains/Cardano/CardanoWalletManager.swift:
```diff
@@ -102,36 +104,146 @@ extension CardanoWalletManager: TransactionSender {
+        }
+
+        let minChange = try minChange()
+        let change = adaBalance - minAdaDecimal
```
Было сложно понять, что тут происходит из-за названия minAdaDecimal. Может minAdaToSendDecimal лучше?

Balashov152 replied:
✅

dbaturin commented on BlockchainSdk/Blockchains/Cardano/CardanoWalletManager.swift:
```diff
@@ -102,36 +104,146 @@ extension CardanoWalletManager: TransactionSender {
+            throw ValidationError.cardanoHasTokens(minimumAmount: minChange)
+        }
+
+        if isTokenBalanceLeft {
+            throw ValidationError.cardanoInsufficientBalanceToSendToken
```
Тут не понятно, почему другая ошибка, а не та, что выше.

Balashov152 replied:
Согласен, упростил, теперь везде `ValidationError.cardanoInsufficientBalanceToSendToken`

dbaturin reviewed (changes requested):
Если честно, разбираться было очень трудно из-за плохой структуры. Ошибки оттуда же растут. Может переделать это, опираясь на реализацию в Андроиде? Там гораздо понятнее всё получилось - https://github.com/tangem/blockchain-sdk-kotlin/pull/611/files#diff-c1d7450a88da5a3dad513ad69528b2ca5f908e5a7740697559609f805e10e79d

ba1ashov committed:
IOS-6286 Simplifying logic

ba1ashov committed:
IOS-6286 Rename minAdaDecimal  -> minAdaToSendDecimal

ba1ashov commented:
> Если честно, разбираться было очень трудно из-за плохой структуры. Ошибки оттуда же растут. Может переделать это, опираясь на реализацию в Андроиде? Там гораздо понятнее всё получилось - https://github.com/tangem/blockchain-sdk-kotlin/pull/611/files#diff-c1d7450a88da5a3dad513ad69528b2ca5f908e5a7740697559609f805e10e79d

Согласен, чет намудрил больше чем нужно. Теперь метод `validateCardanoTokenWithdrawal` более понятный стал, и добавил комментов заодно

ba1ashov committed:
Merge branch 'develop' into IOS-6286-Cardano-tokens

dbaturin commented on BlockchainSdk/Blockchains/Cardano/CardanoWalletManager.swift:
```diff
@@ -102,36 +104,131 @@ extension CardanoWalletManager: TransactionSender {
+        let minChange = try minChange(amount: amount)
+        var change = adaBalance - amount.value
+
+        if amount.type == fee.type {
+            change -= fee.value
```
Не понятно, зачем проверка, ведь мы уже с коинами работаем

Balashov152 replied:
на всякий случай )

dbaturin commented on BlockchainSdk/Blockchains/Cardano/CardanoWalletManager.swift:
```diff
@@ -102,36 +104,131 @@ extension CardanoWalletManager: TransactionSender {
+        let change = adaBalance - minAdaToSendDecimal
+
+        // If there not enough ada balance to change
+        guard change > 0, change < minChange.value else {
+            return
+        }
+
+        throw ValidationError.cardanoInsufficientBalanceToSendToken
```
Странновато выглядит, что throw является дефолтным поведением. С коинами обратная реализация и выглядит гораздо лучше.

dbaturin reviewed (dismissed):
Вроде всё ок, стало заметно проще разбираться

Andoran90 commented on BlockchainSdk/Blockchains/Cardano/CardanoWalletManager.swift:
```diff
@@ -102,36 +104,131 @@ extension CardanoWalletManager: TransactionSender {
+        let change = adaBalance - minAdaToSendDecimal
+
+        // If there not enough ada balance to change
+        guard change > 0, change < minChange.value else {
+            return
+        }
+
+        throw ValidationError.cardanoInsufficientBalanceToSendToken
```
Согласен с Денисом, что странно получается. Давай лучше вот так
```suggestion
        if 0 < change, change < minChange.value {
            throw ValidationError.cardanoInsufficientBalanceToSendToken
        }
```

кстати, может проверка должна быть 
``` swift
    if change < 0 || change < minChange.value {
        throw ValidationError.cardanoInsufficientBalanceToSendToken
    }
```

Balashov152 replied:
Обновил на `if`. Не тут правильная проверка, что сдача "есть" и что она "больше чем минимум"

ba1ashov committed:
IOS-6286 Update CardanoWalletManager.swift

ba1ashov committed:
IOS-6286 Merge branch 'develop' into IOS-6286-Cardano-tokens

ba1ashov committed:
Merge branch 'develop' into IOS-6286-Cardano-tokens

ba1ashov committed:
Merge branch 'develop' into IOS-6286-Cardano-tokens

ba1ashov committed:
IOS-6286 withdrawalSuggestion -> withdrawalNotification

ba1ashov commented:
@tureck1y @Andoran90 @dbaturin @fedorov-d можете плиз поставить, одно [переименование](https://github.com/tangem/blockchain-sdk-swift/pull/666/commits/57670f00fdedc6fe7e11fb8afd166bad9f60a500) все таки решил сделать, что бы все одинаково было
